File: apps/qr-api/internal/adapters/inbound/http/problem/problem.go
// Package problem implements the RFC 9457 contract shared with stats-api; titles fixed per slug.
package problem

import (
	"fmt"
	"time"
)

// URNPrefix is the base of every problem type.
const URNPrefix = "urn:proyectot:problem:"

// Slug identifies a problem type.
type Slug string

// The shared error registry (same slugs and titles as stats-api).
const (
	SlugMalformedJSON               Slug = "malformed-json"
	SlugUnauthorized                Slug = "unauthorized"
	SlugNotFound                    Slug = "not-found"
	SlugMethodNotAllowed            Slug = "method-not-allowed"
	SlugPayloadTooLarge             Slug = "payload-too-large"
	SlugRequestHeaderFieldsTooLarge Slug = "request-header-fields-too-large"
	SlugUnsupportedMediaType        Slug = "unsupported-media-type"
	SlugValidationError             Slug = "validation-error"
	SlugTooManyRequests             Slug = "too-many-requests"
	SlugInternalError               Slug = "internal-error"
	SlugDownstreamUnavailable       Slug = "downstream-unavailable"
	SlugDownstreamTimeout           Slug = "downstream-timeout"
)

// registry maps each slug to its status and its exact title; clients branch on type, not title.
var registry = map[Slug]struct {
	Title  string
	Status int
}{
	SlugMalformedJSON:               {"Malformed JSON body", 400},
	SlugUnauthorized:                {"Authentication required", 401},
	SlugNotFound:                    {"Resource not found", 404},
	SlugMethodNotAllowed:            {"Method not allowed", 405},
	SlugPayloadTooLarge:             {"Payload too large", 413},
	SlugRequestHeaderFieldsTooLarge: {"Request header fields too large", 431},
	SlugUnsupportedMediaType:        {"Unsupported media type", 415},
	SlugValidationError:             {"The request body failed validation", 422},
	SlugTooManyRequests:             {"Too many requests", 429},
	SlugInternalError:               {"Internal server error", 500},
	SlugDownstreamUnavailable:       {"Upstream dependency unavailable", 502},
	SlugDownstreamTimeout:           {"Upstream dependency timed out", 504},
}

// Title returns the exact title registered for a slug.
func Title(s Slug) string { return registry[s].Title }

// Status returns the HTTP status registered for a slug.
func Status(s Slug) int { return registry[s].Status }

// Issue is one entry of `errors[]` on a validation problem.
type Issue struct {
	Pointer string `json:"pointer"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

// Document is the serialized problem, in the exact shape of the contract.
type Document struct {
	Type      string  `json:"type"`
	Title     string  `json:"title"`
	Status    int     `json:"status"`
	Detail    string  `json:"detail,omitempty"`
	Instance  string  `json:"instance"`
	RequestID string  `json:"requestId"`
	Timestamp string  `json:"timestamp"`
	Errors    []Issue `json:"errors,omitempty"`
}

// Error is the in-flight problem: handlers return it, the Fiber error handler renders it.
type Error struct {
	Slug   Slug
	Detail string
	Issues []Issue
	// TokenPresented adds error="invalid_token"; RFC 6750 §3 allows it only for a sent token.
	TokenPresented bool
	// Retry-After value in seconds, set on 429.
	RetryAfterSeconds int
	// cause is kept for the log only; it never reaches the client.
	cause error
}

func (e *Error) Error() string {
	if e.Detail != "" {
		return fmt.Sprintf("%s: %s", e.Slug, e.Detail)
	}
	return string(e.Slug)
}

// Unwrap exposes the original error to errors.Is/As and to the logger.
func (e *Error) Unwrap() error { return e.cause }

// WithTokenPresented marks the 401 as a rejected token; it changes WWW-Authenticate only.
func (e *Error) WithTokenPresented() *Error {
	e.TokenPresented = true
	return e
}

// Status returns the HTTP status of the problem.
func (e *Error) Status() int { return Status(e.Slug) }

// Document renders the problem for one request.
func (e *Error) Document(instance, requestID string, now time.Time) Document {
	return Document{
		Type:      URNPrefix + string(e.Slug),
		Title:     Title(e.Slug),
		Status:    Status(e.Slug),
		Detail:    e.Detail,
		Instance:  instance,
		RequestID: requestID,
		Timestamp: now.UTC().Format("2006-01-02T15:04:05.000Z"),
		Errors:    e.Issues,
	}
}

// New builds a problem with the default detail of its slug.
func New(slug Slug, detail string) *Error { return &Error{Slug: slug, Detail: detail} }

// Wrap builds a problem that keeps the underlying cause for the log.
func Wrap(slug Slug, detail string, cause error) *Error {
	return &Error{Slug: slug, Detail: detail, cause: cause}
}

// Constructors for the contract's fixed documents; details match contracts/examples.

// MalformedJSON is the 400 of an unparseable body.
func MalformedJSON(cause error) *Error {
	return Wrap(SlugMalformedJSON, "The request body is not valid JSON.", cause)
}

// Unauthorized is the single 401 document; tokenPresented changes the header, not the body.
func Unauthorized(tokenPresented bool) *Error {
	return &Error{
		Slug:           SlugUnauthorized,
		Detail:         "The access token is missing, expired or invalid.",
		TokenPresented: tokenPresented,
	}
}

// NotFound is the 404 of an unknown path.
func NotFound() *Error {
	return New(SlugNotFound, "The requested resource does not exist.")
}

// MethodNotAllowed is the 405 of a known path with the wrong method.
func MethodNotAllowed() *Error {
	return New(SlugMethodNotAllowed, "This method is not allowed on the requested resource.")
}

// PayloadTooLarge is the 413 of a body above MAX_BODY_BYTES.
func PayloadTooLarge(limit int) *Error {
	return New(SlugPayloadTooLarge, fmt.Sprintf("The request body exceeds the %d byte limit.", limit))
}

// RequestHeaderFieldsTooLarge is the 431 of a request line or header block above the buffer.
func RequestHeaderFieldsTooLarge() *Error {
	return New(SlugRequestHeaderFieldsTooLarge, "The request headers exceed the server's buffer.")
}

// UnsupportedMediaType is the 415 of a non-JSON body.
func UnsupportedMediaType() *Error {
	return New(SlugUnsupportedMediaType, "Content-Type must be application/json.")
}

// Validation is the 422 carrying every failed rule.
func Validation(detail string, issues []Issue) *Error {
	return &Error{Slug: SlugValidationError, Detail: detail, Issues: issues}
}

// TooManyRequests is the 429 of the rate limiter.
func TooManyRequests(retryAfterSeconds int) *Error {
	return &Error{
		Slug:              SlugTooManyRequests,
		Detail:            "Rate limit exceeded. Retry after the window resets.",
		RetryAfterSeconds: retryAfterSeconds,
	}
}

// Internal is the 500: no details ever leave the process.
func Internal(cause error) *Error {
	return Wrap(SlugInternalError, "An unexpected error occurred. Quote the requestId when reporting it.", cause)
}

// DownstreamUnavailable is the 502 of a failed call to stats-api.
func DownstreamUnavailable(cause error) *Error {
	return Wrap(SlugDownstreamUnavailable, "The statistics service did not return a usable response.", cause)
}

// DownstreamTimeout is the 504 of a call to stats-api that ran out of time.
func DownstreamTimeout(cause error) *Error {
	return Wrap(SlugDownstreamTimeout, "The statistics service did not respond in time.", cause)
}

File: apps/stats-api/src/adapters/inbound/http/problem.ts
import type { Request, Response } from 'express';
import { requestIdOf } from './middleware/request-id.middleware.js';

/** Registry of `contracts/problems.md` §2; titles are copied verbatim. */
export const PROBLEM_TYPE_PREFIX = 'urn:proyectot:problem:';

export type ProblemSlug =
  | 'malformed-json'
  | 'unauthorized'
  | 'not-found'
  | 'method-not-allowed'
  | 'payload-too-large'
  | 'request-header-fields-too-large'
  | 'unsupported-media-type'
  | 'validation-error'
  | 'too-many-requests'
  | 'internal-error';

interface ProblemDefinition {
  readonly status: number;
  readonly title: string;
  readonly detail: string;
}

export const PROBLEM_REGISTRY: Readonly<Record<ProblemSlug, ProblemDefinition>> = {
  'malformed-json': {
    status: 400,
    title: 'Malformed JSON body',
    detail: 'The request body is not valid JSON.',
  },
  unauthorized: {
    status: 401,
    title: 'Authentication required',
    detail: 'The access token is missing, expired or invalid.',
  },
  'not-found': {
    status: 404,
    title: 'Resource not found',
    detail: 'The requested resource does not exist.',
  },
  'method-not-allowed': {
    status: 405,
    title: 'Method not allowed',
    detail: 'This method is not supported on this resource.',
  },
  'payload-too-large': {
    status: 413,
    title: 'Payload too large',
    detail: 'The request body is larger than the configured limit.',
  },
  'request-header-fields-too-large': {
    status: 431,
    title: 'Request header fields too large',
    detail: "The request headers exceed the server's buffer.",
  },
  'unsupported-media-type': {
    status: 415,
    title: 'Unsupported media type',
    detail: 'Content-Type must be application/json.',
  },
  'validation-error': {
    status: 422,
    title: 'The request body failed validation',
    detail: 'The request body failed validation.',
  },
  'too-many-requests': {
    status: 429,
    title: 'Too many requests',
    detail: 'Rate limit exceeded. Retry after the window resets.',
  },
  'internal-error': {
    status: 500,
    title: 'Internal server error',
    detail: 'An unexpected error occurred. Quote the requestId when reporting it.',
  },
};

/** `errors[].code` — the shared enum of `contracts/problems.md` §3. */
export type ValidationCode =
  | 'missing_field'
  | 'invalid_type'
  | 'empty_matrix'
  | 'empty_row'
  | 'ragged_row'
  | 'non_finite_value'
  | 'too_many_rows'
  | 'too_many_cols'
  | 'too_many_matrices'
  | 'too_many_elements'
  | 'numerical_overflow';

export interface ValidationIssue {
  readonly pointer: string;
  readonly code: ValidationCode;
  readonly message: string;
}

export interface ProblemDocument {
  readonly type: string;
  readonly title: string;
  readonly status: number;
  readonly detail: string;
  readonly instance: string;
  readonly requestId: string;
  readonly timestamp: string;
  readonly errors?: readonly ValidationIssue[];
}

export interface ProblemOptions {
  readonly detail?: string;
  readonly issues?: readonly ValidationIssue[];
  readonly headers?: Readonly<Record<string, string>>;
}

/** A failure that already knows its problem document; the error middleware renders it. */
export class ProblemError extends Error {
  public readonly slug: ProblemSlug;
  public readonly options: ProblemOptions;

  public constructor(slug: ProblemSlug, options: ProblemOptions = {}) {
    super(options.detail ?? PROBLEM_REGISTRY[slug].detail);
    this.name = 'ProblemError';
    this.slug = slug;
    this.options = options;
  }
}

/** Request path without the query string — the `instance` member of the document. */
export function instanceOf(req: Request): string {
  const [path] = req.originalUrl.split('?');
  return path === undefined || path.length === 0 ? req.originalUrl : path;
}

export function buildProblem(
  req: Request,
  res: Response,
  slug: ProblemSlug,
  options: ProblemOptions = {},
): ProblemDocument {
  const definition = PROBLEM_REGISTRY[slug];
  const base = {
    type: `${PROBLEM_TYPE_PREFIX}${slug}`,
    title: definition.title,
    status: definition.status,
    detail: options.detail ?? definition.detail,
    instance: instanceOf(req),
    requestId: requestIdOf(res),
    timestamp: new Date().toISOString(),
  };
  return options.issues === undefined ? base : { ...base, errors: options.issues };
}

// Buffer, not string: `res.send` appends `; charset=utf-8`, and the contract fixes the media type
// at exactly `application/problem+json`.
export function sendProblem(
  req: Request,
  res: Response,
  slug: ProblemSlug,
  options: ProblemOptions = {},
): void {
  const document = buildProblem(req, res, slug, options);
  for (const [name, value] of Object.entries(options.headers ?? {})) {
    res.setHeader(name, value);
  }
  res.setHeader('Content-Type', 'application/problem+json');
  res.status(document.status).send(Buffer.from(JSON.stringify(document), 'utf8'));
}
